TST: Test division by zero both with scalar and with array

ckastner opened the pull request:
These two approaches may produce different results. See #15562, where on
armv7l, the latter approach did not raise as expected.

ckastner committed:
TST: Check zero division both with scalar and with array

These two approaches may produce different results. See #15562, where on
armv7l, the latter approach did not raise as expected.

seberg commented:
This seems right, but it seems like the next step will be adding a skipif to the whole test? Can you check that and do that? With an explanation that this seems to be a platform/cpu bug (or missing feature).

seberg commented:
In some cases paths are different due to SIMD usage I believe, i.e. it is possible the scalar path uses different cpu instructions to do the same job.

ckastner committed:
TST: Skip test known to fail on ARM with soft-float

ckastner committed:
TST: Update out-of-date skip condition

The value for platform.machine() was reported from Debian in 2012, but current
Debian systems no longer this value. Switch the test to the new arm_softfloat
flag, which is the actual underlying cause of the test failure.

ckastner commented:
Unfortunately there doesn't seem to be an clear-cut way to determine whether a specific implementation uses floating point emulation or a hardware FPU. It's possible for a system to have a hardware FPU and still use soft-float, if compiled for such.

I checked sysconfig.get_config_vars() for candidate information, and picked HOST_GNU_TYPE as the most likely candidate to succeed. If it's undefined, everything is as it was before. If it's defined and ends with 'gnueabi' (as opposed to eg 'gnueabihf'), the test is skipped.

This works for Debian systems. I cannot guarantee that it will work for all affected systems, but so far, it seems that only Debian reported this, and in any case, the test can easily be modified or extended in future.

ckastner commented:
I waited for the initial tests to complete, and now merged the typo commit into another to reduce the noise.

seberg commented on numpy/core/tests/test_errstate.py:
```diff
@@ -1,12 +1,18 @@
 
 class TestErrstate:
-    @pytest.mark.skipif(platform.machine() == "armv5tel", reason="See gh-413.")
+    @pytest.mark.skipif(arm_softfloat, reason="See gh-413, gh-15562.")
```
Maybe you can make the reason longer?

ckastner replied:
> Maybe you can make the reason longer?

Based on your previous suggestion, I 'platform/cpu issue with FPU'. I can make it longer, of course, but looking at the other tests, brevity seems to be preferred.

seberg reviewed (approved):
Thanks, I am in general happy with this. Does anyone think that inspecting the `HOST_GNU_TYPE ` env variable is not reasonable here/in the test?

mattip commented:
Since the goal is to skip tests on a buggy platform, which should only happen infrequently, I am fine with this method of discovery.

seberg commented:
Thanks for the followups and care with figuring out how to blacklist the test!

ckastner commented:
Thank you for the review1
